src/Http/Routing/src/Internal/MapActionExpressionTreeBuilder.cs

```diff
@@ -0,0 +1,402 @@
+                    var queryProperty = Expression.Property(HttpRequestExpr, nameof(HttpRequest.Query));
+                    paramterExpression = BindParamenter(queryProperty, parameter, queryAttribute.Name);
+                }
+                else if (parameter.GetCustomAttributes().OfType<IFromHeaderMetadata>().FirstOrDefault() is { } headerAttribute)
+                {
+                    var headersProperty = Expression.Property(HttpRequestExpr, nameof(HttpRequest.Headers));
+                    paramterExpression = BindParamenter(headersProperty, parameter, headerAttribute.Name);
+                }
+                else if (parameter.GetCustomAttributes().OfType<IFromBodyMetadata>().FirstOrDefault() is { } bodyAttribute)
+                {
+                    if (consumeBodyDirectly)
+                    {
+                        throw new InvalidOperationException("Action cannot have more than one FromBody attribute.");
+                    }
+
+                    if (consumeBodyAsForm)
+                    {
+                        ThrowCannotReadBodyDirectlyAndAsForm();
+                    }
+
+                    consumeBodyDirectly = true;
+                    allowEmptyBody = bodyAttribute.AllowEmpty;
+                    bodyType = parameter.ParameterType;
+                    paramterExpression = Expression.Convert(DeserializedBodyArg, bodyType);
+                }
+                else if (parameter.GetCustomAttributes().OfType<IFromFormMetadata>().FirstOrDefault() is { } formAttribute)
+                {
+                    if (consumeBodyDirectly)
+                    {
+                        ThrowCannotReadBodyDirectlyAndAsForm();
+                    }
+
+                    consumeBodyAsForm = true;
+
+                    var formProperty = Expression.Property(HttpRequestExpr, nameof(HttpRequest.Form));
+                    paramterExpression = BindParamenter(formProperty, parameter, parameter.Name);
+                }
+                else if (parameter.CustomAttributes.Any(a => typeof(IFromServiceMetadata).IsAssignableFrom(a.AttributeType)))
+                {
+                    paramterExpression = Expression.Call(GetRequiredServiceMethodInfo.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr);
+                }
+                else
+                {
+                    if (parameter.ParameterType == typeof(IFormCollection))
+                    {
+                        if (consumeBodyDirectly)
+                        {
+                            ThrowCannotReadBodyDirectlyAndAsForm();
+                        }
+
+                        consumeBodyAsForm = true;
+
+                        paramterExpression = Expression.Property(HttpRequestExpr, nameof(HttpRequest.Form));
+                    }
+                    else if (parameter.ParameterType == typeof(HttpContext))
+                    {
+                        paramterExpression = HttpContextParameter;
+                    }
```

Comment:
Is there a plan to support cancellation token?
```suggestion
                    }
                    else if (parameter.ParameterType == typeof(CancellationToken))
```

Reply:
Seems reasonable. I assume you'd want it to be the HttpContext.RequestAborted token. @davidfowl @pranavkm @javiercn Thoughts?

Reply:
That's what MVC binds it to today.

Reply:
@halter73 do you accept PR for MapAction yet? If so I would like to send one to support `CancellationToken`.